style: Less ID revalidation selectors.

emilio opened the pull request:
Avoid adding id selectors that are in the rule hash keyed by that ID to the list
of revalidation selectors.

This partially fixes bug 1369611 (we could look at the rule hash itself to avoid
inserting some more into the list of revalidation selectors).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17966)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/stylist.rs
 * @canaltinova: components/style/stylist.rs

emilio commented:
r? @bzbarsky 

(I don't really know why you wrote that `#foo1 > span` shouldn't be a revalidation selector, so I removed it)

bzbarsky commented:
> I don't really know why you wrote that `#foo1 > span` shouldn't be a revalidation selector

Because `foo1` is an id in the rulehash.  Though maybe I misunderstood the test?  Is the test actually setting up a stylist and putting the selectors in there?  Or is it calling needs_revalidation on selectors in isolation, without the full context of a stylist?

As in, what you're doing here is a partial fix for https://bugzilla.mozilla.org/show_bug.cgi?id=1369611 not a full fix.  It's good, but we can perhaps do better with the full thing.  Separate issue on that, of course.  You should probably leave the comment about how we can do better and the reference to the bug about it, though.

bzbarsky commented:
Oh, and the PR should not claim to fix bug 1369611, since it doesn't.

With the comment fixes, and if the test just tests selectors individually, r=me, I guess, though we'll need a better test once we fix bug 1369611. 

Also, don't you still need to remove the first FIXME comment, and move its bug link to the `#foo1 > span` case?

emilio commented:
> Because foo1 is an id in the rulehash. Though maybe I misunderstood the test? Is the test actually setting up a stylist and putting the selectors in there? Or is it calling needs_revalidation on selectors in isolation, without the full context of a stylist?

Oh, ok, I see what you mean now. Yeah, it's calling `needs_revalidation` without the context of a stylist. Also, it's not clear to me how to get it perfectly without doing multiple passes, that is, setting all the selectors in the rule hash, then scanning the revalidation selectors. Though perhaps just checking what we already have is good enough.

bzbarsky commented:
Yeah, it very well might be, so we might wontfix bug 1369611.  Will need to do some measuring.

emilio committed:
style: Less ID revalidation selectors.

Avoid adding id selectors that are in the rule hash keyed by that ID to the list
of revalidation selectors.

This partially fixes bug 1369611 (we could look at the rule hash itself to avoid
inserting some more into the list of revalidation selectors).

emilio commented:
@bors-servo r=bzbarsky

bors-servo commented:
:pushpin: Commit f746756 has been approved by `bzbarsky`

<!-- @bors-servo r=bzbarsky f74675639d615429ef5600353e993324e04cf9e4 -->

bors-servo commented:
:hourglass: Testing commit f74675639d615429ef5600353e993324e04cf9e4 with merge 16704bbaaa9ec324972788f1a455639d4bbef0fd...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8205), [arm32](http://build.servo.org/builders/arm32/builds/8428), [arm64](http://build.servo.org/builders/arm64/builds/8461), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8547), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5310), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5343), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8544), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1020), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1023), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5245), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5303), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1022), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1026), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3786)
Approved by: bzbarsky
Pushing 16704bbaaa9ec324972788f1a455639d4bbef0fd to master...
